internal/juju/applications.go

```diff
@@ -179,131 +256,147 @@ func resolveCharmURL(charmName string) (*charm.URL, error) {
 }
 
 func (c applicationsClient) CreateApplication(ctx context.Context, input *CreateApplicationInput) (*CreateApplicationResponse, error) {
-	appName := input.ApplicationName
-	if appName == "" {
-		appName = input.CharmName
+	conn, err := c.GetConnection(&input.ModelName)
+	if err != nil {
+		return nil, err
 	}
-	if err := names.ValidateApplicationName(appName); err != nil {
+	defer func() { _ = conn.Close() }()
+
+	transformedInput, err := input.validateAndTransform()
+	if err != nil {
 		return nil, err
 	}
 
-	conn, err := c.GetConnection(&input.ModelName)
+	applicationAPIClient := apiapplication.NewClient(conn)
+	if applicationAPIClient.BestAPIVersion() >= 19 {
+		err = c.deployFromRepository(applicationAPIClient, transformedInput)
+	} else {
+		err = c.legacyDeploy(ctx, conn, applicationAPIClient, transformedInput)
+		err = jujuerrors.Annotate(err, "legacy deploy method")
+	}
 	if err != nil {
 		return nil, err
 	}
-	defer func() { _ = conn.Close() }()
 
+	// If we have managed to deploy something, now we have
+	// to check if we have to expose something
+	err = c.processExpose(applicationAPIClient, transformedInput.applicationName, transformedInput.expose)
+
+	return &CreateApplicationResponse{
+		AppName: transformedInput.applicationName,
+	}, err
+}
+
+func (c applicationsClient) deployFromRepository(applicationAPIClient *apiapplication.Client, transformedInput transformedCreateApplicationInput) error {
+	settingsForYaml := map[interface{}]interface{}{transformedInput.applicationName: transformedInput.config}
+	configYaml, err := goyaml.Marshal(settingsForYaml)
+	if err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	c.Tracef("Calling DeployFromRepository")
+	_, _, errs := applicationAPIClient.DeployFromRepository(apiapplication.DeployFromRepositoryArg{
+		CharmName:       transformedInput.charmName,
+		ApplicationName: transformedInput.applicationName,
+		Base:            &transformedInput.charmBase,
+		Channel:         &transformedInput.charmChannel,
+		ConfigYAML:      string(configYaml),
+		Cons:            transformedInput.constraints,
+		NumUnits:        &transformedInput.units,
+		Placement:       transformedInput.placement,
+		Revision:        &transformedInput.charmRevision,
+		Trust:           transformedInput.trust,
+	})
+	return errors.Join(errs...)
+}
+
+func (c applicationsClient) legacyDeploy(ctx context.Context, conn api.Connection, applicationAPIClient *apiapplication.Client, transformedInput transformedCreateApplicationInput) error {
```

Comment:
Maybe it's worth commenting on how long we should not drop this legacy?

Reply:
agreed